### ISSUE TITLE
SubClass annotations are missing from the base class

### ISSUE THREAD

**Issue by yaelah** [swagger-api/swagger-codegen#2449]
In order for jackson ObjectMapper to deserialize correctly a base class to its subclasses, we need to include annotations defining all expected subclasses in the parent class.
The codegen currently does not generate those annotations.
For this example YAML:

definitions:
&nbsp;&nbsp;Base:
&nbsp;&nbsp;&nbsp;&nbsp;type: object
&nbsp;&nbsp;&nbsp;&nbsp;discriminator: disc
&nbsp;&nbsp;&nbsp;&nbsp;properties:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;disc:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;type: string
&nbsp;&nbsp;&nbsp;&nbsp;required:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;\- disc
&nbsp;&nbsp;SubClass:
&nbsp;&nbsp;&nbsp;&nbsp;allOf:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;\- $ref: '#/definitions/Base'
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;required:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;\- field
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;properties:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;field:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;type: string

Jackson expects these annotations to be produced, but they are not:
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "disc")
@JsonSubTypes({ @Type(value = SubClass.class, name = "SubClass") })
public class Base {
}

**Comment by wing328** [comment]
@yaelah thanks for the detailed above. Would you have cycle to contribute a fix?

Here is a good starting point: https://github.com/swagger-api/swagger-codegen/blob/master/modules/swagger-codegen/src/main/resources/Java/model.mustache

**Comment by wing328** [comment]
Similar ticket: https://github.com/swagger-api/swagger-codegen/issues/1253

**Comment by yaelah** [comment]
I have a fix for this. What is the process for submitting patches? I have never submitted anything to this project before.
thanks.

**Comment by wing328** [comment]
Please refer to https://github.com/swagger-api/swagger-codegen/blob/master/CONTRIBUTING.md for contributing guidelines.

If you need help, just reply to let us know.

**Comment by yaelah** [comment]
Thanks for noticing my wrong email.

**Comment by alex-estela** [comment]
Hi guys,
Also currently facing this issue. Any news about the release of this fix ?
Thanks in advance

**Comment by triptec** [comment]
@wing328 Do you know anything about progress on this?

**Comment by wing328** [comment]
I believe @yaelah has fixed the issue in his branch but didn't have time to submit a PR yet.

Can you test his fix via https://github.com/yaelah/swagger-codegen to see if it resolves the issue for you?

If the fix works, I can cherry-pick the commits to submit a PR for the fix.

**Comment by szakrewsky** [comment]
I modified @yaelah post to work on the latest master branch.  I have been using this for a while now, hope it helps.  One issue I did notice is that it won't apply to sub-sub classes.

**Comment by wing328** [comment]
@szakrewsky I added a comment to the PR #4085. Please take a look when you've time.

**Comment by jeff9finger** [comment]
+1

**Comment by jeff9finger** [comment]
How is the discriminator value determined for the name property of the`@JsonSubTypes.@Type` annotation. It looks like the best that can be done without more information is to use the class name.

But if a discriminator property value for a sub class is not the name of the class, but another string, how can that be determined? This may need to be addressed in a new issue, but simply wanted to raise it here to see if anyone has  ideas.

My idea is to add a codegen vendor extension that can be used to state the discriminator value in a subclass - maybe `x-java-discriminator-value`

```
definitions:
  Base:
    type: object
    discriminator: disc
    properties:
      disc:
        type: string
    required:
      - disc
  SubClass:
    x-java-discriminator-value: sub-class
    allOf:
      - $ref: '#/definitions/Base'
        required:
          - field
        properties:
          field:
            type: string 
```

**Comment by szakrewsky** [comment]
@jeff9finger I have seen a similar issue. The spec (http://swagger.io/specification/) for the discriminator field says
> When used, the value MUST be the name of this schema or any schema that inherits it.

And since the class name is the schema name its all good.

**Comment by jeff9finger** [comment]
Using the classname is a good default (in fact, it looks to be the only thing that makes sense). But in some cases, the discriminator value needs to be something other than the class name.

I have used a vendor extension "x-discriminator-value" on each subclass. This value contains the value from the enumeration defined for that subclass. Changed the template to handle that value if it exists (otherwise, use name as before). This works perfectly for my. I'd  like to see it in the master branch so I don't have to have my own templates.... Anyone else having this issue?
